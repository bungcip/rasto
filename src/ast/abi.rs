//! The `ast` module contains the definitions for the Abstract Syntax Tree (AST) nodes
//! that represent Rust ABIs.

use crate::pretty_printer::{PrettyPrinter, Printer};
use std::fmt;

/// Represents a Rust ABI (Application Binary Interface).
///
/// This enum is used to specify the calling convention for functions,
/// especially in the context of `extern` blocks and function pointers.
///
/// # Examples
///
/// ```rust
/// use rasto::ast::Abi;
///
/// // A named ABI, like "C"
/// let c_abi = Abi::Named("C".to_string());
///
/// // A named ABI, like "system"
/// let system_abi = Abi::Named("system".to_string());
/// ```
#[derive(Debug, Clone, PartialEq)]
pub enum Abi {
    /// Represents a named ABI, specified as a string literal.
    ///
    /// For example, `"C"`, `"system"`, or `"Rust"`.
    Named(String),
}

impl PrettyPrinter for Abi {
    fn pretty_print<'a>(&'a self, printer: &mut Printer<'a>) -> fmt::Result {
        match self {
            Abi::Named(name) => {
                printer.string("\"");
                printer.string(name);
                printer.string("\"");
            }
        }
        Ok(())
    }
}
