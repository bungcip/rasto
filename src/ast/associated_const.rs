//! An associated constant in a trait or impl.

use crate::ast::expressions::Expr;
use crate::ast::ident::Ident;
use crate::ast::metadata::Md;
use crate::ast::types::Type;

/// Represents an associated constant in a trait or implementation.
///
/// An associated constant is a constant value that is associated with a specific
/// type. It is defined within a trait or an `impl` block.
///
/// # Examples
///
/// In a trait definition:
///
/// ```rust
/// trait MyTrait {
///     const MAX: u16;
/// }
/// ```
///
/// In an `impl` block:
///
/// ```rust
/// trait MyTrait {
///     const MAX: u16;
/// }
/// struct MyType;
///
/// impl MyTrait for MyType {
///     const MAX: u16 = 123;
/// }
/// ```
#[derive(Debug, Clone, PartialEq)]
pub struct AssociatedConst {
    /// The name of the associated constant.
    pub ident: Ident,
    /// The type of the associated constant.
    pub ty: Type,
    /// The optional expression that provides the value of the constant.
    ///
    /// This is `None` in a trait definition where the value is not specified,
    /// and `Some` in an `impl` block where the value is provided.
    pub expr: Option<Box<Expr>>,
    /// Metadata associated with the constant, such as attributes and comments.
    pub md: Option<Box<Md>>,
}
