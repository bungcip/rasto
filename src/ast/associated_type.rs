use crate::ast::generics::GenericParams;
use crate::ast::metadata::Md;
use crate::ast::types::Type;
use thin_vec::ThinVec;

/// An associated type in a trait.
#[derive(Debug, Clone, PartialEq)]
pub struct AssociatedType {
    /// The name of the associated type.
    pub ident: String,
    /// The generic parameters of the associated type.
    pub generics: GenericParams,
    /// The bounds on the associated type.
    pub bounds: ThinVec<Type>,
    /// The default type for the associated type.
    pub default: Option<Type>,
    /// Metadata about the associated type.
    pub md: Option<Box<Md>>,
}
