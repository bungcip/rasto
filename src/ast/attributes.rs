//! Defines the AST nodes for attributes.
//!
//! Attributes are metadata that can be attached to various items in Rust code. They are
//! enclosed in `#[...]` for outer attributes and `#![...]` for inner attributes.

use crate::ast::literals::Lit;
use crate::pretty_printer::{PrettyPrinter, Printer};
use std::fmt;
use thin_vec::ThinVec;

/// An attribute, such as `#[repr(C)]` or `#![allow(dead_code)]`.
#[derive(Debug, Clone, PartialEq)]
pub enum Attribute {
    /// An inner attribute, such as `#![allow(dead_code)]`.
    Inner(Meta),
    /// An outer attribute, such as `#[repr(C)]`.
    Outer(Meta),
}

/// The meta item within an attribute.
///
/// For example, in `#[repr(C)]`, the meta item is `repr(C)`.
#[derive(Debug, Clone, PartialEq)]
pub enum Meta {
    /// A meta list, such as `repr(C)`.
    List(MetaList),
    /// A meta path, such as `test`.
    Path(String),
    /// A meta name-value pair, such as `key = "value"`.
    NameValue(MetaNameValue),
}

/// A meta list, such as `repr(C)`.
#[derive(Debug, Clone, PartialEq)]
pub struct MetaList {
    /// The path of the meta list, e.g., `repr`.
    pub path: String,
    /// The meta items within the list, e.g., `C`.
    pub metas: ThinVec<Meta>,
}

/// A meta name-value pair, such as `key = "value"`.
#[derive(Debug, Clone, PartialEq)]
pub struct MetaNameValue {
    /// The path of the meta name-value pair, e.g., `key`.
    pub path: String,
    /// The value of the meta name-value pair, e.g., `"value"`.
    pub value: Lit,
}

impl PrettyPrinter for Attribute {
    fn pretty_print<'a>(&'a self, printer: &mut Printer<'a>) -> fmt::Result {
        match self {
            Attribute::Inner(meta) => {
                printer.string("#![");
                meta.pretty_print(printer)?;
                printer.string("]");
            }
            Attribute::Outer(meta) => {
                printer.string("#[");
                meta.pretty_print(printer)?;
                printer.string("]");
            }
        }
        Ok(())
    }
}

impl PrettyPrinter for Meta {
    fn pretty_print<'a>(&'a self, printer: &mut Printer<'a>) -> fmt::Result {
        match self {
            Meta::List(list) => list.pretty_print(printer),
            Meta::Path(path) => {
                printer.string(path);
                Ok(())
            }
            Meta::NameValue(name_value) => name_value.pretty_print(printer),
        }
    }
}

impl PrettyPrinter for MetaList {
    fn pretty_print<'a>(&'a self, printer: &mut Printer<'a>) -> fmt::Result {
        printer.string(&self.path);
        printer.string("(");
        for (i, meta) in self.metas.iter().enumerate() {
            if i > 0 {
                printer.string(", ");
            }
            meta.pretty_print(printer)?;
        }
        printer.string(")");
        Ok(())
    }
}

impl PrettyPrinter for MetaNameValue {
    fn pretty_print<'a>(&'a self, printer: &mut Printer<'a>) -> fmt::Result {
        printer.string(&self.path);
        printer.string(" = ");
        self.value.pretty_print(printer)
    }
}
