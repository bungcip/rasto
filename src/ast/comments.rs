//! Defines the representation of comments in the AST.

use compact_str::CompactString;

/// Represents a comment in the source code.
///
/// Comments can be either line comments or block comments.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Comment {
    /// A line comment, starting with `//`.
    ///
    /// The string contains the content of the comment, without the `//`.
    ///
    /// # Example
    ///
    /// ```text
    /// // This is a line comment.
    /// ```
    Line(CompactString),
    /// A block comment, enclosed in `/* ... */`.
    ///
    /// The string contains the content of the comment, without the `/*` and `*/`.
    ///
    /// # Example
    ///
    /// ```text
    /// /* This is a block comment. */
    /// ```
    Block(CompactString),
    /// A doc comment, starting with `///`.
    ///
    /// The string contains the content of the comment, without the `///`.
    ///
    /// # Example
    ///
    /// ```text
    /// /// This is a doc comment.
    /// ```
    Doc(CompactString),
}
