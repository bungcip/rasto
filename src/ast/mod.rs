//! The `ast` module contains the definitions for the Abstract Syntax Tree (AST) nodes.
//!
//! This module defines the data structures that represent the syntactic structure of Rust code.
//! It includes nodes for expressions, statements, items, literals, and more. The AST is used
//! by the pretty-printer to generate formatted Rust code.

#[macro_use]
mod macros;

/// Defines the AST node for an `extern` block.
pub mod abi;
/// Defines the AST node for an associated const in a trait.
pub mod associated_const;
/// Defines the AST node for an associated type in a trait.
pub mod associated_type;
/// Defines the AST nodes for attributes.
pub mod attributes;
/// Defines the AST nodes for comments.
pub mod comments;
/// Defines the AST nodes for expressions.
pub mod expressions;
/// Defines the AST node for a file.
pub mod file;
/// Defines the AST nodes for generics.
pub mod generics;
/// Defines the AST node for an `asm!` expression.
pub mod item_asm;
/// Defines the AST node for a `const`, `static`, or `type` item.
pub mod item_def;
/// Defines the AST node for an enum.
pub mod item_enum;
/// Defines the AST node for an `extern` block.
pub mod item_extern_block;
/// Defines the AST node for an `extern crate` item.
pub mod item_extern_crate;
/// Defines the AST node for a function.
pub mod item_fn;
/// Defines the AST node for a foreign module.
pub mod item_foreign_mod;
/// Defines the AST node for an `impl` item.
pub mod item_impl;
/// Defines the AST node for a macro definition.
pub mod item_macro;
/// Defines the AST node for a module.
pub mod item_mod;
/// Defines the AST node for a static item.
pub mod item_static;
/// Defines the AST node for a struct.
pub mod item_struct;
/// Defines the AST node for a trait.
pub mod item_trait;
/// Defines the AST node for a trait alias.
pub mod item_trait_alias;
/// Defines the AST node for a `union` item.
pub mod item_union;
/// Defines the AST node for a `use` item.
pub mod item_use;
/// Defines the AST nodes for items.
pub mod items;
/// Defines the AST nodes for keywords.
pub mod keyword;
/// Defines the AST nodes for literals.
pub mod literals;
/// Defines the AST nodes for metadata.
pub mod metadata;
/// Defines the AST nodes for patterns.
pub mod patterns;
/// Defines the AST nodes for statements.
pub mod statements;
/// Defines the AST nodes for tokens.
pub mod tokens;
/// Defines the AST nodes for types.
pub mod types;
/// Defines the AST nodes for visibility.
pub mod visibility;
/// Defines the AST nodes for `where` clauses.
pub mod where_clause;

pub use abi::*;
pub use associated_const::*;
pub use associated_type::*;
pub use attributes::*;
pub use comments::*;
pub use expressions::*;
pub use file::*;
pub use generics::*;
pub use item_asm::*;
pub use item_def::*;
pub use item_enum::*;
pub use item_extern_block::*;
pub use item_extern_crate::*;
pub use item_fn::*;
pub use item_foreign_mod::*;
pub use item_impl::*;
pub use item_macro::*;
pub use item_mod::*;
pub use item_static::*;
pub use item_struct::*;
pub use item_trait::*;
pub use item_trait_alias::*;
pub use item_union::*;
pub use item_use::*;
pub use literals::*;
pub use metadata::*;
pub use patterns::*;
pub use statements::*;
pub use tokens::*;
pub use types::*;
pub use visibility::*;
pub use where_clause::*;
