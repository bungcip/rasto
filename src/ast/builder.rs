use crate::ast::*;

pub fn file() -> FileBuilder {
    FileBuilder::new()
}

pub struct FileBuilder {
    items: Vec<Item>,
}

impl FileBuilder {
    pub fn new() -> Self {
        Self { items: vec![] }
    }

    pub fn item(mut self, item: impl Into<Item>) -> Self {
        self.items.push(item.into());
        self
    }

    pub fn build(self) -> File {
        File { items: self.items }
    }
}

pub fn fn_def(name: impl Into<String>) -> FnBuilder {
    FnBuilder::new(name)
}

pub struct FnBuilder {
    ident: String,
    block: Option<Block>,
}

impl FnBuilder {
    pub fn new(name: impl Into<String>) -> Self {
        Self {
            ident: name.into(),
            block: None,
        }
    }

    pub fn block(mut self, block: Block) -> Self {
        self.block = Some(block);
        self
    }

    pub fn build(self) -> ItemFn {
        let block = self.block.expect("block is required");

        ItemFn {
            leading_comments: vec![],
            sig: Signature { ident: self.ident },
            block,
            trailing_comments: vec![],
        }
    }
}

pub fn expr() -> ExprBuilder {
    ExprBuilder
}

#[derive(Clone, Copy)]
pub struct ExprBuilder;

impl ExprBuilder {
    pub fn array(self, elems: impl IntoIterator<Item = Expr>) -> Expr {
        Expr::Array(ExprArray {
            elems: elems.into_iter().collect(),
        })
    }

    pub fn assign(self, left: Expr, right: Expr) -> Expr {
        Expr::Assign(ExprAssign {
            left: Box::new(left),
            right: Box::new(right),
        })
    }

    pub fn async_block(self, block: Block) -> Expr {
        Expr::Async(ExprAsync { block })
    }

    pub fn await_expr(self, expr: Expr) -> Expr {
        Expr::Await(ExprAwait {
            expr: Box::new(expr),
        })
    }

    pub fn binary(self, left: Expr, op: BinOp, right: Expr) -> Expr {
        Expr::Binary(ExprBinary {
            left: Box::new(left),
            op,
            right: Box::new(right),
        })
    }

    pub fn block(self, block: Block) -> Expr {
        Expr::Block(ExprBlock { block })
    }

    pub fn break_expr(self) -> Expr {
        Expr::Break(ExprBreak)
    }

    pub fn call(self, func: Expr, args: impl IntoIterator<Item = Expr>) -> Expr {
        Expr::Call(ExprCall {
            func: Box::new(func),
            args: args.into_iter().collect(),
        })
    }

    pub fn cast(self, expr: Expr, ty: impl Into<String>) -> Expr {
        Expr::Cast(ExprCast {
            expr: Box::new(expr),
            ty: ty.into(),
        })
    }

    pub fn closure(
        self,
        inputs: impl IntoIterator<Item = impl Into<String>>,
        body: Expr,
    ) -> Expr {
        Expr::Closure(ExprClosure {
            inputs: inputs.into_iter().map(Into::into).collect(),
            body: Box::new(body),
        })
    }

    pub fn const_block(self, block: Block) -> Expr {
        Expr::Const(ExprConst { block })
    }

    pub fn continue_expr(self) -> Expr {
        Expr::Continue(ExprContinue)
    }

    pub fn field(self, expr: Expr, member: impl Into<String>) -> Expr {
        Expr::Field(ExprField {
            expr: Box::new(expr),
            member: member.into(),
        })
    }

    pub fn for_loop(self, pat: impl Into<String>, expr: Expr, body: Block) -> Expr {
        Expr::For(ExprFor {
            pat: pat.into(),
            expr: Box::new(expr),
            body,
        })
    }

    pub fn if_expr(self, cond: Expr, then_branch: Block, else_branch: Option<Expr>) -> Expr {
        Expr::If(ExprIf {
            cond: Box::new(cond),
            then_branch,
            else_branch: else_branch.map(Box::new),
        })
    }

    pub fn index(self, expr: Expr, index: Expr) -> Expr {
        Expr::Index(ExprIndex {
            expr: Box::new(expr),
            index: Box::new(index),
        })
    }

    pub fn lit(self, lit: impl Into<Lit>) -> Expr {
        Expr::Lit(lit.into())
    }

    pub fn loop_expr(self, body: Block) -> Expr {
        Expr::Loop(ExprLoop { body })
    }

    pub fn macro_call(self, ident: impl Into<String>, tokens: TokenStream) -> Expr {
        Expr::MacroCall(ExprMacroCall {
            ident: ident.into(),
            tokens,
        })
    }

    pub fn match_expr(self, expr: Expr, arms: impl IntoIterator<Item = Arm>) -> Expr {
        Expr::Match(ExprMatch {
            expr: Box::new(expr),
            arms: arms.into_iter().collect(),
        })
    }

    pub fn method_call(
        self,
        receiver: Expr,
        method: impl Into<String>,
        args: impl IntoIterator<Item = Expr>,
    ) -> Expr {
        Expr::MethodCall(ExprMethodCall {
            receiver: Box::new(receiver),
            method: method.into(),
            args: args.into_iter().collect(),
        })
    }

    pub fn paren(self, expr: Expr) -> Expr {
        Expr::Paren(ExprParen {
            expr: Box::new(expr),
        })
    }

    pub fn range(self, start: Option<Expr>, limits: RangeLimits, end: Option<Expr>) -> Expr {
        Expr::Range(ExprRange {
            start: start.map(Box::new),
            limits,
            end: end.map(Box::new),
        })
    }

    pub fn reference(self, is_mut: bool, expr: Expr) -> Expr {
        Expr::Reference(ExprRef {
            is_mut,
            expr: Box::new(expr),
        })
    }

    pub fn return_expr(self, expr: Option<Expr>) -> Expr {
        Expr::Return(ExprReturn {
            expr: expr.map(Box::new),
        })
    }

    pub fn struct_expr(
        self,
        path: impl Into<String>,
        fields: impl IntoIterator<Item = FieldValue>,
    ) -> Expr {
        Expr::Struct(ExprStruct {
            path: path.into(),
            fields: fields.into_iter().collect(),
        })
    }

    pub fn tuple(self, elems: impl IntoIterator<Item = Expr>) -> Expr {
        Expr::Tuple(ExprTuple {
            elems: elems.into_iter().collect(),
        })
    }

    pub fn while_loop(self, cond: Expr, body: Block) -> Expr {
        Expr::While(ExprWhile {
            cond: Box::new(cond),
            body,
        })
    }
}

pub fn const_item(name: impl Into<String>, ty: impl Into<String>, expr: impl Into<Expr>) -> ItemConstBuilder {
    ItemConstBuilder::new(name, ty, expr)
}

pub struct ItemConstBuilder {
    ident: String,
    ty: String,
    expr: Box<Expr>,
    leading_comments: Vec<Comment>,
    trailing_comments: Vec<Comment>,
}

impl ItemConstBuilder {
    pub fn new(name: impl Into<String>, ty: impl Into<String>, expr: impl Into<Expr>) -> Self {
        Self {
            ident: name.into(),
            ty: ty.into(),
            expr: Box::new(expr.into()),
            leading_comments: vec![],
            trailing_comments: vec![],
        }
    }

    pub fn leading_comment(mut self, comment: impl Into<Comment>) -> Self {
        self.leading_comments.push(comment.into());
        self
    }

    pub fn trailing_comment(mut self, comment: impl Into<Comment>) -> Self {
        self.trailing_comments.push(comment.into());
        self
    }

    pub fn build(self) -> ItemConst {
        ItemConst {
            ident: self.ident,
            ty: self.ty,
            expr: self.expr,
            leading_comments: self.leading_comments,
            trailing_comments: self.trailing_comments,
        }
    }
}

pub fn extern_crate_item(name: impl Into<String>) -> ItemExternCrateBuilder {
    ItemExternCrateBuilder::new(name)
}

pub struct ItemExternCrateBuilder {
    ident: String,
    leading_comments: Vec<Comment>,
    trailing_comments: Vec<Comment>,
}

impl ItemExternCrateBuilder {
    pub fn new(name: impl Into<String>) -> Self {
        Self {
            ident: name.into(),
            leading_comments: vec![],
            trailing_comments: vec![],
        }
    }

    pub fn leading_comment(mut self, comment: impl Into<Comment>) -> Self {
        self.leading_comments.push(comment.into());
        self
    }

    pub fn trailing_comment(mut self, comment: impl Into<Comment>) -> Self {
        self.trailing_comments.push(comment.into());
        self
    }

    pub fn build(self) -> ItemExternCrate {
        ItemExternCrate {
            ident: self.ident,
            leading_comments: self.leading_comments,
            trailing_comments: self.trailing_comments,
        }
    }
}

pub fn foreign_mod_item(abi: impl Into<String>) -> ItemForeignModBuilder {
    ItemForeignModBuilder::new(abi)
}

pub struct ItemForeignModBuilder {
    abi: String,
    items: Vec<Item>,
    leading_comments: Vec<Comment>,
    trailing_comments: Vec<Comment>,
}

impl ItemForeignModBuilder {
    pub fn new(abi: impl Into<String>) -> Self {
        Self {
            abi: abi.into(),
            items: vec![],
            leading_comments: vec![],
            trailing_comments: vec![],
        }
    }

    pub fn item(mut self, item: impl Into<Item>) -> Self {
        self.items.push(item.into());
        self
    }

    pub fn leading_comment(mut self, comment: impl Into<Comment>) -> Self {
        self.leading_comments.push(comment.into());
        self
    }

    pub fn trailing_comment(mut self, comment: impl Into<Comment>) -> Self {
        self.trailing_comments.push(comment.into());
        self
    }

    pub fn build(self) -> ItemForeignMod {
        ItemForeignMod {
            abi: self.abi,
            items: self.items,
            leading_comments: self.leading_comments,
            trailing_comments: self.trailing_comments,
        }
    }
}

pub fn macro_item(expr: impl Into<Expr>) -> ItemMacroBuilder {
    ItemMacroBuilder::new(expr)
}

pub struct ItemMacroBuilder {
    expr: Expr,
    leading_comments: Vec<Comment>,
    trailing_comments: Vec<Comment>,
}

impl ItemMacroBuilder {
    pub fn new(expr: impl Into<Expr>) -> Self {
        Self {
            expr: expr.into(),
            leading_comments: vec![],
            trailing_comments: vec![],
        }
    }

    pub fn leading_comment(mut self, comment: impl Into<Comment>) -> Self {
        self.leading_comments.push(comment.into());
        self
    }

    pub fn trailing_comment(mut self, comment: impl Into<Comment>) -> Self {
        self.trailing_comments.push(comment.into());
        self
    }

    pub fn build(self) -> ItemMacro {
        ItemMacro {
            expr: self.expr,
            leading_comments: self.leading_comments,
            trailing_comments: self.trailing_comments,
        }
    }
}

pub fn mod_item(name: impl Into<String>) -> ItemModBuilder {
    ItemModBuilder::new(name)
}

pub struct ItemModBuilder {
    ident: String,
    content: Option<Vec<Item>>,
    leading_comments: Vec<Comment>,
    trailing_comments: Vec<Comment>,
}

impl ItemModBuilder {
    pub fn new(name: impl Into<String>) -> Self {
        Self {
            ident: name.into(),
            content: None,
            leading_comments: vec![],
            trailing_comments: vec![],
        }
    }

    pub fn content(mut self, content: Vec<Item>) -> Self {
        self.content = Some(content);
        self
    }

    pub fn leading_comment(mut self, comment: impl Into<Comment>) -> Self {
        self.leading_comments.push(comment.into());
        self
    }

    pub fn trailing_comment(mut self, comment: impl Into<Comment>) -> Self {
        self.trailing_comments.push(comment.into());
        self
    }

    pub fn build(self) -> ItemMod {
        ItemMod {
            ident: self.ident,
            content: self.content,
            leading_comments: self.leading_comments,
            trailing_comments: self.trailing_comments,
        }
    }
}

pub fn static_item(name: impl Into<String>, ty: impl Into<String>, expr: impl Into<Expr>) -> ItemStaticBuilder {
    ItemStaticBuilder::new(name, ty, expr)
}

pub struct ItemStaticBuilder {
    ident: String,
    ty: String,
    expr: Box<Expr>,
    leading_comments: Vec<Comment>,
    trailing_comments: Vec<Comment>,
}

impl ItemStaticBuilder {
    pub fn new(name: impl Into<String>, ty: impl Into<String>, expr: impl Into<Expr>) -> Self {
        Self {
            ident: name.into(),
            ty: ty.into(),
            expr: Box::new(expr.into()),
            leading_comments: vec![],
            trailing_comments: vec![],
        }
    }

    pub fn leading_comment(mut self, comment: impl Into<Comment>) -> Self {
        self.leading_comments.push(comment.into());
        self
    }

    pub fn trailing_comment(mut self, comment: impl Into<Comment>) -> Self {
        self.trailing_comments.push(comment.into());
        self
    }

    pub fn build(self) -> ItemStatic {
        ItemStatic {
            ident: self.ident,
            ty: self.ty,
            expr: self.expr,
            leading_comments: self.leading_comments,
            trailing_comments: self.trailing_comments,
        }
    }
}

pub fn trait_alias_item(name: impl Into<String>, bounds: Vec<String>) -> ItemTraitAliasBuilder {
    ItemTraitAliasBuilder::new(name, bounds)
}

pub struct ItemTraitAliasBuilder {
    ident: String,
    bounds: Vec<String>,
    leading_comments: Vec<Comment>,
    trailing_comments: Vec<Comment>,
}

impl ItemTraitAliasBuilder {
    pub fn new(name: impl Into<String>, bounds: Vec<String>) -> Self {
        Self {
            ident: name.into(),
            bounds,
            leading_comments: vec![],
            trailing_comments: vec![],
        }
    }

    pub fn leading_comment(mut self, comment: impl Into<Comment>) -> Self {
        self.leading_comments.push(comment.into());
        self
    }

    pub fn trailing_comment(mut self, comment: impl Into<Comment>) -> Self {
        self.trailing_comments.push(comment.into());
        self
    }

    pub fn build(self) -> ItemTraitAlias {
        ItemTraitAlias {
            ident: self.ident,
            bounds: self.bounds,
            leading_comments: self.leading_comments,
            trailing_comments: self.trailing_comments,
        }
    }
}

pub fn type_item(name: impl Into<String>, ty: impl Into<String>) -> ItemTypeBuilder {
    ItemTypeBuilder::new(name, ty)
}

pub struct ItemTypeBuilder {
    ident: String,
    ty: String,
    leading_comments: Vec<Comment>,
    trailing_comments: Vec<Comment>,
}

impl ItemTypeBuilder {
    pub fn new(name: impl Into<String>, ty: impl Into<String>) -> Self {
        Self {
            ident: name.into(),
            ty: ty.into(),
            leading_comments: vec![],
            trailing_comments: vec![],
        }
    }

    pub fn leading_comment(mut self, comment: impl Into<Comment>) -> Self {
        self.leading_comments.push(comment.into());
        self
    }

    pub fn trailing_comment(mut self, comment: impl Into<Comment>) -> Self {
        self.trailing_comments.push(comment.into());
        self
    }

    pub fn build(self) -> ItemType {
        ItemType {
            ident: self.ident,
            ty: self.ty,
            leading_comments: self.leading_comments,
            trailing_comments: self.trailing_comments,
        }
    }
}

pub fn union_item(name: impl Into<String>) -> ItemUnionBuilder {
    ItemUnionBuilder::new(name)
}

pub struct ItemUnionBuilder {
    ident: String,
    fields: Vec<Field>,
    leading_comments: Vec<Comment>,
    trailing_comments: Vec<Comment>,
}

impl ItemUnionBuilder {
    pub fn new(name: impl Into<String>) -> Self {
        Self {
            ident: name.into(),
            fields: vec![],
            leading_comments: vec![],
            trailing_comments: vec![],
        }
    }

    pub fn field(mut self, field: Field) -> Self {
        self.fields.push(field);
        self
    }

    pub fn leading_comment(mut self, comment: impl Into<Comment>) -> Self {
        self.leading_comments.push(comment.into());
        self
    }

    pub fn trailing_comment(mut self, comment: impl Into<Comment>) -> Self {
        self.trailing_comments.push(comment.into());
        self
    }

    pub fn build(self) -> ItemUnion {
        ItemUnion {
            ident: self.ident,
            fields: self.fields,
            leading_comments: self.leading_comments,
            trailing_comments: self.trailing_comments,
        }
    }
}

pub fn use_item(path: impl Into<String>) -> ItemUseBuilder {
    ItemUseBuilder::new(path)
}

pub struct ItemUseBuilder {
    path: String,
    leading_comments: Vec<Comment>,
    trailing_comments: Vec<Comment>,
}

impl ItemUseBuilder {
    pub fn new(path: impl Into<String>) -> Self {
        Self {
            path: path.into(),
            leading_comments: vec![],
            trailing_comments: vec![],
        }
    }

    pub fn leading_comment(mut self, comment: impl Into<Comment>) -> Self {
        self.leading_comments.push(comment.into());
        self
    }

    pub fn trailing_comment(mut self, comment: impl Into<Comment>) -> Self {
        self.trailing_comments.push(comment.into());
        self
    }

    pub fn build(self) -> ItemUse {
        ItemUse {
            path: self.path,
            leading_comments: self.leading_comments,
            trailing_comments: self.trailing_comments,
        }
    }
}
