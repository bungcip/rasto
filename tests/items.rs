use rasto::ast::{
    AssociatedType, Item, TokenStream, Type,
    builder::{
        const_kind, def_item, extern_crate_item, fn_def, foreign_mod_item, macro_item, mod_item,
        static_kind, trait_alias_item, trait_def, type_alias_kind, union_item, use_item,
    },
    expr,
};
use thin_vec::thin_vec;

#[test]
fn test_def_item() {
    let const_item = def_item("MAX", const_kind(Type::from("u16"), expr().lit(234342))).build();
    insta::assert_snapshot!(const_item.to_string());

    let static_item = def_item("COUNTER", static_kind(Type::from("u32"), expr().lit(0))).build();
    insta::assert_snapshot!(static_item.to_string());

    let type_item = def_item(
        "MyResult<T>",
        type_alias_kind(Type::from("Result<T, MyError>")),
    )
    .build();
    insta::assert_snapshot!(type_item.to_string());
}

#[test]
fn test_extern_crate_item() {
    let item = extern_crate_item("serde").build();
    insta::assert_snapshot!(item.to_string());
}

use rasto::ast::builder::block;

#[test]
fn test_foreign_mod_item() {
    let item = foreign_mod_item("C")
        .item(Item::Fn(fn_def("foo").block(block()).build()))
        .build();
    insta::assert_snapshot!(item.to_string());
}

#[test]
fn test_macro_item() {
    let item = macro_item(expr().macro_call(
        "my_macro",
        TokenStream {
            tokens: thin_vec![],
        },
    ))
    .build();
    insta::assert_snapshot!(item.to_string());
}

#[test]
fn test_mod_item() {
    let item = mod_item("my_module").build();
    insta::assert_snapshot!(item.to_string());
}

#[test]
fn test_mod_item_with_content() {
    let item = mod_item("my_module")
        .content(thin_vec![Item::Fn(fn_def("foo").block(block()).build())])
        .build();
    insta::assert_snapshot!(item.to_string());
}

#[test]
fn test_trait_alias_item() {
    let item = trait_alias_item(
        "ShareableIterator",
        thin_vec!["Iterator".to_string(), "Sync".to_string()],
    )
    .build();
    insta::assert_snapshot!(item.to_string());
}

#[test]
fn test_union_item() {
    let item = union_item("MyUnion")
        .field("f1", Type::from("u32"))
        .field("f2", Type::from("f32"))
        .build();
    insta::assert_snapshot!(item.to_string());
}

#[test]
fn test_use_item() {
    let item = use_item("std::collections::HashMap").build();
    insta::assert_snapshot!(item.to_string());
}

#[test]
fn test_trait_with_associated_type() {
    let item = trait_def("MyTrait").associated_type(AssociatedType {
        ident: "MyType".to_string(),
        generics: Default::default(),
        bounds: thin_vec![],
        default: None,
        md: None,
    });
    insta::assert_snapshot!(item.build().to_string());
}
