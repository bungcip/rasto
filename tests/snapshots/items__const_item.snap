---
source: tests/items.rs
expression: item.to_string()
---
const MAX: u16 = 234342;
